aws-android-sdk-pinpoint/src/main/java/com/amazonaws/mobileconnectors/pinpoint/analytics/AnalyticsClient.java

```diff
@@ -359,33 +359,38 @@ public void removeGlobalMetric(String eventType, String metricName) {
     }
 
     /**
-     * Adds the specified campaign attributes to events to track Campaign Analytic
-     * <p>
-     * You should not use this method as it will be called by the NotificationManager when the app is opened
-     * from a push notification.
-     *
-     * @param campaign the map with campaign attributes of the campaign received
+     * Replaces the current event source attributes (if any)
+     * with a new set of event source attributes
+     * @param attributes map of attribute values
      */
-    public void setCampaignAttributes(Map<String, String> campaign) {
-        if (campaign == null) {
-            log.warn("Null campaign attributes provided to setCampaignAttributes.");
-            return;
+    public void updateEventSourceGlobally(final Map<String, String> attributes) {
+        //This call removes previous events identifiers from the
+        //globalAttributes collection in AnalyticsEvent.
+        clearEventSourceAttributes();
+        for (final Map.Entry<String, String> entry : attributes.entrySet()) {
+            if (entry.getValue() != null) {
+                addGlobalAttribute(entry.getKey(), entry.getValue());
+            }
         }
-
-        campaignAttributes = campaign;
+        //We set this so the analytics client knows what to remove from
+        //from the globalAttributes collection when it's time to clear it
+        this.eventSourceAttributes = attributes;
```

Comment:
If `attributes` is _not_ a `ConcurrentHashMap<>()`, you'll lose your assignment type, that you'd chose earlier.

It may be better to make the class field as `final`, and then use `eventSourceAttributes.addAll(attributes);`, instead.

Reply:
Ah...good point. Updating that in the next commit.